keps/sig-network/3070-reserved-service-ip-range/README.md

```diff
@@ -0,0 +1,751 @@
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+Kubernetes Services are an abstract way to expose an application running on a
+set of Pods. Services has a ClusterIP that is virtual and allows to load-balance
+traffic accross the different Pods. This ClusterIP can be assigned:
+- dynamically, the cluster will pick one within the configured Service IP range
+- statically, the user will set one IP within the configured Service IP range
+
+Currently, there is no possibility, when creating a Service with a static Cluster IP
+to know if the IP has been chosen randomly by any other Service using dynamic allocation.
+
+Defining a predefined range in the Service IP subnet to not allocate dynamic address from it,
+allow to avoid any possible conflict with Services with static IP inside that range.
+
+
+## Motivation
+
+There are situation that users need to rely on well-known predefined IP addresses.
+
+The best example is the Kubernetes DNS, the IP is hardcoded in the kubelet flag
+`--cluster-dns strings`. For historical reasons and by convention the Service IP
+for the DNS use to the the 10th IP from the range, however, until the DNS Service
+has been created this IP can be taken by chance by any Service created with a dynamically
+allocated IP.
+
+There are also other use cases like the one explained in 
+https://github.com/kubernetes/kubernetes/issues/95570
+
+
+### Goals
+
+- allow to create a Service with a static IPs without any risk of IP conflict.
+
+### Non-Goals
+
+- changing the allocators implementation (bitmaps, ...).
+- add new configuration options.
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+"As a developer I want to be able to create Services with well-known static IPs with the guarantee
+that those IPs will not be hijacked by a Service using dynamic IP allocation"
```

Comment:
That's not a user story, that's a restatement of the Summary with the words "I want" before it.

The user stories here are (1) DNS getting the 10th IP so you can pass it to `--cluster-dns`, and (2) the use case from https://github.com/kubernetes/kubernetes/issues/95570